react/the_react_ecosystem/fetching-data.md

```diff
@@ -0,0 +1,297 @@
+
+> Reminder: Whenever you are changing state, (or doing anything else that could cause a rerender) inside of useEffect, be sure to always pass a dependency. If you forget to do so useEffect will enter an infinite loop and rerender exponentially.
+
+You might be wondering what's going on here
+
+~~~jsx
+return (
+  imageURL && (
+    <>
+      <h1>An image</h1>
+      <img src={imageURL alt={"placeholder text"}}>
+    </>
+  )
+)
+~~~
+
+This is a pattern called _short circuiting_ and it allows developers to conditionally render JSX without using if statements or ternarys. The above snippet can be read as, "if imageURL is truthy then continue to evaluate the expression". A truthy value and an object evaluates to the object so in this case, JSX is returned. If imageURL was falsy, the right hand side of the && operation wouldn't evaluate and the component would return false.
+
+### Handling errors
+
+Working over the network is inherently unreliable. The API you're making a request to might be down, there could be network connectivity issues, or the response you recieve could contain errors. Any number of things can go wrong and if you don't preemptively plan for errors, your website can break or appear unresponsive to users. To simulate a network error, scroll up to that previous code snippet and change the fetch url to something random. The page will remain a white screen without giving the user any indication that the page has finished loading or that there was an error.
+
+To fix this, we need to check _something_ in the return of the Image component. We'll call it "error".
+
+~~~jsx
+  return (
+    <>
+      {error && <p>A network error was encountered</p>}
+      {imageURL && (
+        <>
+          <h1>An image</h1>
+          <img src={imageURL} alt={"placeholder text"} />
+        </>
+      )}
+    </>
+  );
+~~~
+
+To set this error, we'll add it to the component's state.
+
+~~~jsx
+const [imageURL, setImageURL] = useState(null);
+const [error, setError] = useState(null);
+~~~
+
+And finally, to give error a value when a request fails, we'll add a conditional to check the response status and set it where our console.error line was.
+
+~~~jsx
+useEffect(() => {
+  fetch("https://jsonplaceholder.typicode.com/photos", { mode: "cors" })
+    .then((response) => {
+      if (response.status >= 400) {
+        throw new Error("server error");
+      }
+      return response.json();
+    })
+    .then((response) => setImageURL(response[0].url))
+    .catch((error) => setError(error));
+}, []);
+~~~
+
+> Note: A fetch request can return an http error without triggering the catch block. Always be sure to check your response and manually throw an error
```

Comment:
This is something new I learned. Can you share a reference link so that I can read more on this?

Reply:
Certainly https://plainenglish.io/blog/handling-unsuccessful-calls-of-fetch-api-7eef551b74c8

It might be more accurate to say "A fetch request can return an http error **_status_** without triggering the catch block".  Catch will "catch" network errors and invalid URLs  but not > 400 status codes